x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitter.java

```diff
@@ -57,9 +57,13 @@ public void process(String[] row, Runnable incrementTrainingDocs, Runnable incre
             throw new IllegalStateException("Unknown class [" + classValue + "]; expected one of " + classSamples.keySet());
         }
 
+        // We use ensure the target sample count is at least 1 as if the cardinality
+        // is too low we might get a target of zero and, thus, no samples of the whole class
+        double targetSampleCount = Math.max(1.0, samplingRatio * sample.cardinality);
```

Comment:
Is it guaranteed that `sample.cardinality` is at least `1`? If it was `0`, then `targetSampleCount` would be `1` but maybe that's not what we'd wanted?

Reply:
Wouldn't it be easier to just take the first value found?

`boolean isTraining = random.nextDouble() <= p || sample.observed == 0;`

Reply:
@przemekwitek If cardinality was zero then we'd never encounter a doc with that class.

However, @benwtrent 's suggestion also works. I'll change it to that as it seems simpler indeed.

Reply:
Actually, as soon as I made that change the unit test we have for checking we uniformly pick samples failed. As this solution means we always pick the first value we see, we introduce a bias towards that. I'll stick to my original solution as it prevents that bias.

Reply:
@dimitris-athanasiou I thought it was not possible to even run a classification job if cardinality was `< 2`. The test talks about a cardinality of `1`.

Reply:
It makes no difference from the perspective of the code but I'll update to use cardinality of `2` for the sake of reflecting actual conditions.

Reply:
Also, note these are different cardinalities we're talking about. The requirement of `2` refers to "there should be at least 2 classes". Whereas the addressed problem arises when the cardinality of the docs belonging to a certain class is low.

Reply:
> Whereas the addressed problem arises when the cardinality of the docs belonging to a certain class is low.

AH! This is confusing 🤔 . The solution as posed makes sense.

Reply:
I have pushed a commit that should make the unit test clearer and more realistic.